---
title: 'Лабораторна робота №6. Функції'
description:
   Документ зроблено за допомогою [Quarto](https://quarto.org/)
lang: ukr
format:
  html:
    code-fold: true
    toc: true # меню
    toc_float: # спливаюче меню  
      collapsed: true # авто
      number_sections: true
jupyter: python3
---

__Мета:__ _навчитися писати функції та користуватися ними_

::: callout-note
## Примітка
Попередньо потрібно опрацювати лекцію 6.
:::

::: callout-note
## Примітка
У лабораторній роботі використано матеріали [python institute.](https://pythoninstitute.org/pcep) 
:::

### Завдання 1

Написати і протестувати функцію, яка приймає один аргумент (рік) і повертає `True`, якщо рік є високосним, або `False` інакше.

Основу функції вже закладено у коді в редакторі.

::: callout-note
## Примітка
Для вас підготовлено короткий тестуючий код, який можна використовувати для перевірки своєї функції.
:::

У коді використовуються два списки - один із тестовими даними, інший - з очікуваними результатами. Код повідомить Вам, якщо якісь із Ваших результатів неправильні.

In [ ]:
def is_year_leap(year):
	return (year % 4 == 0 and (year % 100 != 0 or year % 400 == 0)) 

test_data = [1900, 2000, 2016, 1987]
test_results = [False, True, True, False]
for i in range(len(test_data)):
	yr = test_data[i]
	print(yr,"->",end=" ")
	result = is_year_leap(yr)
	print(result)
	if result == test_results[i]:
		print("OK")
	else:
		print("Failed")

1900 -> False
OK
2000 -> True
OK
2016 -> True
OK
1987 -> False
OK


### Завдання 2 

Написати та протестувати функцію, яка приймає два аргументи (рік і місяць) і повертає кількість днів для даної пари рік-місяць (у той час, як тільки лютий чутливий до значення `year`, Ваша функція має бути універсальною).

Початок функції готовий. Тепер переконайте функцію повернути `None`, якщо її аргументи не мають сенсу.

Звичайно, Ви можете (і повинні) використовувати раніше написану та протестовану функцію (див. попередні лаб. роб.). Це може бути дуже корисним. Ми рекомендуємо використовувати список із зазначенням довжини місяців. Ви можете створити його всередині функції – цей трюк значно скоротить код.

Ми підготували тестуючий код. Розвійте його, щоб увімкнути більше тестових випадків.

In [17]:
def is_year_leap(year):
  return (year % 4 == 0 and (year % 100 != 0 or year % 400 == 0))

def days_in_month(year, month):
  if month < 1 or month > 12:
    return None
  if year < 1:
    return None
  
  days_in_months = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]

  if month == 2:
    if is_year_leap(year):
      return 29
    else:
      return 28
    
  return days_in_months[month - 1]

test_years = [1900, 2000, 2016, 1987]
test_months = [2, 2, 1, 11]
test_results = [28, 29, 31, 30]
for i in range(len(test_years)):
  yr = test_years[i]
  mo = test_months[i]
  print(yr, mo, "->", end=" ")
  result = days_in_month(yr, mo)
  if result == test_results[i]:
    print("OK")
  else:
    print("Failed")


1900 2 -> OK
2000 2 -> OK
2016 1 -> OK
1987 11 -> OK


### Завдання 3

Написати та протестувати функцію, яка приймає три аргументи (рік, місяць та день місяця) та повертає відповідний день року або None, якщо якийсь із аргументів невірний.

Використовуйте раніше написані та протестовані функції. Додайте до коду свої власні тестові випадки.

In [27]:
def is_year_leap(year):
  return (year % 4 == 0 and (year % 100 != 0 or year % 400 == 0))

def days_in_month(year, month):
  if month < 1 or month > 12:
    return None
  if year < 1:
    return None
  
  days_in_months = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]

  if month == 2:
    if is_year_leap(year):
      return 29
    else:
      return 28
    
  return days_in_months[month - 1]

def day_of_year(year, month, day):
  if year < 1 or month < 1 or month > 12 or day < 1:
    return None
  
  days_in_this_month = days_in_month(year, month)

  if day > days_in_this_month:
    return None

  day_of_year = 0
  for i in range(1, month):
    day_of_year += days_in_month(year, i)

  day_of_year += day

  return day_of_year
  
print(day_of_year(2000, 12, 31))

366


###  Завдання 4

Натуральне число є простим, якщо воно більше 1 і не має дільників, крім 1 і самого себе.

Важко? Зовсім ні. Наприклад, 8 не є простим числом, оскільки Ви можете розділити його на 2 та 4 (ми не можемо використовувати дільники, рівні 1 та 8, оскільки визначення забороняє це).

З іншого боку, 7 - просте число, оскільки ми можемо знайти йому підходящих дільників.


Ваше завдання - написати функцію, що перевіряє, чи є число простим чи ні.

Функція:

* називається `is_prime`;
* приймає один аргумент (значення для перевірки)
* повертає `True`, якщо аргумент є простим числом, і `False` інакше.  

::: callout-note
## Підказка
Спробуйте розділити аргумент на всі наступні значення (починаючи з 2) і перевірте залишок - якщо він дорівнює нулю, Ваше число не може бути простим; добре подумайте, коли вам слід зупинити процес.
:::

Якщо вам потрібно знайти квадратний корінь із будь-якого значення, Ви можете використовувати оператор `**`. Пам'ятайте: квадратний корінь `x` дорівнює $x^{0,5}$.

Доповніть код у редакторі.

In [31]:
def is_prime(num):
  if num < 2:
    return False
  
  for i in range(2, int(num**0.5) + 1):
    if num % i == 0:
      return False
  
  return True

for i in range(1, 20):
  if is_prime(i + 1):
    print(i + 1, end=" ")
print()

2 3 5 7 11 13 17 19 


### Очікуваний вивод

`2 3 5 7 11 13 17 19`

### Завдання 5

Витрата палива автомобіля можна висловити по-різному. Наприклад, у Європі він відображається як кількість витраченого палива на 100 кілометрів шляху.

У США він відображається як кількість миль, пройдених автомобілем під час витрачання одного галону палива.

Ваше завдання - написати дві функції, що конвертують л/100км в миль на галон, і навпаки.

Функції:

* названі `liters_100km_to_miles_gallon` та `miles_gallon_to_liters_100km` відповідно;
* приймають один аргумент (значення, що відповідає їхнім іменам)  

Доповніть код у редакторі.

Запустіть свій код і перевірте, чи Ваш результат збігається з нашим.

Ось деяка інформація, яка може допомогти Вам:

* 1 американська мідя = 1609.344 метрів;
* 1 американський галон = 3.785411784 літрів.

In [40]:
def liters_100km_to_miles_gallon(liters):
  return ((100 / 1.609344) / (liters / 3.785411784))

def miles_gallon_to_liters_100km(miles):
  return ((100 * 3.785411784) / (miles * 1.609344))

print(liters_100km_to_miles_gallon(3.9))
print(liters_100km_to_miles_gallon(7.5))
print(liters_100km_to_miles_gallon(10.))
print(miles_gallon_to_liters_100km(60.3))
print(miles_gallon_to_liters_100km(31.4))
print(miles_gallon_to_liters_100km(23.5))

60.31143162393162
31.36194444444444
23.52145833333333
3.9007393587617467
7.490910297239915
10.009131205673757


### Очікуваний вивод

`60.31143162393162`  
`31.36194444444444`  
`23.52145833333333`  
`3.9007393587617467`  
`7.490910297239916`  
`10.009131205673757`

### Завдання 6 

Написати функцію, яка перевіряє, чи три сторони заданої довжини побудувати трикутник. Вона має повернути `True`, якщо сторони можуть утворити трикутник, і `False` в іншому випадку. У цьому випадку `is_a_triangle(a, b, c)` – гарна назва для такої функції.

In [51]:
def is_a_triangle(a, b, c):
  if a + b > c and a + c > b and b + c > a:
    return True
  else:
    return False

print(is_a_triangle(5, 20, 7))

False


### Завдання 7 

Написати функцію `is_a_right_triangle(a, b, c)`, яка перевіряє, чи є трикутник зі сторонами `a`, `b`, `c` прямокутним. При цьому потрібно використати функцію `is_a_triangle(a, b, c)` з завдання 6. 

In [50]:
def is_a_triangle(a, b, c):
  if a + b > c and a + c > b and b + c > a:
    return True
  else:
    return False
  
def is_a_right_triangle(a, b, c):
  if not is_a_triangle(a, b, c):
    return False
  
  s = sorted([a, b, c])
  a1 = s[0]
  b1 = s[1]
  c1 = s[2]

  if c1**2 == a1**2 + b1**2:
    return True
  else:
    return False
  
print(is_a_right_triangle(5, 3, 4))

True


## Завдання для самостіної роботи

1. Виконати завдання 1-7 наведені вище у цьому зошиті.

1. Створити файл __lab_6_StudentLastName.py__ з написаним кодом. 

1. Закомітити файл у локальний репозиторій.

1. Відправити ("запушити") поточну версію Git-проєкта у віддалений репозиторій на GitHub.

1. Звіт має складатися з файлу (за основу взяти __цей Python-зошит__)  `lab_6_StudentLastName.ipynb`. (Можливі якісь додакові файли)

## Контрольні запитання

1. Який вивід наступного фрагмента?  


  На консоль виведе помилку, тому що змінна alt невизначена, вона тільки доступна всередині функції message().

2. Який вивід наступного фрагмента?  


  На консоль вивід буде наступним:
  
  2
  
  1

  Значення a було визначено як 1, тоді як у функції її значення було змінено, але через те що в тілі функції, зміни ніяк не впливають по за нею, тому не вивело два однакових значення.

3. Який вивід наступного фрагмента?  


  На консоль вивід буде наступним:
  
  2
  
  3

  Ключове слово global у Python дозволяє змінювати глобальні змінні у функції, тому після виклику функції вивело значення, яке було змінено в тілі функції.

4. Який вивід наступного фрагмента?  


  На консоль вивід буде наступним:
  
  2
  
  2

  Через те що виклик функції перевизначив змінну a на значення 2, яка знаходиться в тілі функції, тому вивід змінної a через функцію виводу, виведе значення, те яке в тілі функції fun().


## References

1. [GitHub](https://github.com/loldominoc/Aleksandrov_project)